### PR TITLE
bindesc: Add read API

### PR DESCRIPTION
**Pull request by yonsch**
This PR adds a bindesc read API, that allows reading and parsing of binary descriptors. This is useful mostly for reading the descriptors of another image on the same device - e.g. bootloader->app, app->bootloader or app0->app1; though it can also be used by an image on itself to walk over its own descriptors.
This idea has been submitted as part of the initial bindesc PR, but was removed because it raised security concerns of an app reading a bootloader's data. This revised proposal tries to solve this issue (and others) by providing three backends for reading the descriptors - memory mapped flash (original), flash, and RAM. This allows the bootloader to copy its descriptors to a shared memory section with the app. The app can then parse the descriptors from RAM, without accessing the bootloader's ROM directly.

### CODE REVIEW

**Review by kartben — CHANGES_REQUESTED** [review]
Looking good! Did an initial pass on some documentation issues (don't forget you can check the output at https://builds.zephyrproject.io/zephyr/pr/66731/docs/services/binary_descriptors/index.html)

File: doc/services/binary_descriptors/index.rst [hunk 1]
```diff
@@ -113,6 +113,44 @@ name is the Kconfig name (with ``CONFIG_BINDESC_`` removed) in lower case. For e
...
+   This option uses the least amount of RAM, but will not work if the flash is not memory mapped,
+   and is not recommended to read a bootloader's descriptors for security concerns.
+3. Flash - Using an internal buffer, the descriptors are read one by one using the flash API,
+   and given to the user while they're in the buffer.
+
+To enable reading descriptors, enable ``CONFIG_BINDESC_READ``. The three backends are
+enabled by these Kconfig symbols, respectively: ``CONFIG_BINDESC_READ_RAM``,
+``CONFIG_BINDESC_READ_MEMORY_MAPPED_FLASH`` and ``CONFIG_BINDESC_READ_FLASH``.
+
+To read the descriptors, a handle to the descriptors should first be initialized:
```
[the remark this answers is not in the document]
Comment (yonsch) [reply doc/services/binary_descriptors/index.rst]: Right, can you recommend an rst formatter that works with Zephyr?

[the remark this answers is not in the document]
Comment (kartben) [reply doc/services/binary_descriptors/index.rst]: I think VS code has a built-in one that seems to do the job whenever I need it (i.e it's at least smart enough to leave code blocks intact, and things like that), sorry that I don't have better recommendations :/

**Comment by pdgendt** [comment]
@yonsch please rebase

**Comment by yonsch** [comment]
@pdgendt I'm currently away so I couldn't attend to this issue. Once I get back I'll reopen it!

File: tests/subsys/bindesc/reading/testcase.yaml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  platform_allow:
+    - native_posix
+    - native_sim
+  integration_platforms:
+    - native_sim
+  tags: bindesc
+tests:
+  bindesc.read:
+    platform_allow:
+      - native_posix
```
Comment (pdgendt) [inline tests/subsys/bindesc/reading/testcase.yaml:11]: `native_posix` shouldn't be added in tests, I think

File: scripts/west_commands/bindesc.py [hunk 3]
```diff
@@ -186,6 +193,15 @@ def custom_search(self, args):
         custom_tag = self.bindesc_gen_tag(custom_type, int(args.id, 16))
         self.common_search(args, custom_tag)
 
+    def get_offset(self, args):
+        image = self.get_image_data(args.file)
+
+        magic = struct.pack('>Q' if self.is_big_endian else 'Q', self.MAGIC)
+        index = image.find(magic)
```
Comment (pdgendt) [inline scripts/west_commands/bindesc.py:200]: What if the magic value would occur multiple times?

File: subsys/bindesc/Kconfig [hunk 4]
```diff
@@ -20,6 +20,15 @@ source "subsys/bindesc/Kconfig.version"
 source "subsys/bindesc/Kconfig.build_time"
 source "subsys/bindesc/Kconfig.host_info"
 
+config BINDESC_DEFINE_MAX_DATA_SIZE
+	int "Bindesc max data size"
+	range 4 65535
```
Comment (pdgendt) [inline subsys/bindesc/Kconfig:25]: Use `$(UINT16_MAX)`

File: subsys/bindesc/Kconfig [hunk 5]
```diff
@@ -31,4 +31,46 @@ config BINDESC_DEFINE_MAX_DATA_SIZE
...
+	bool "Bindesc read from flash"
+	help
+	  Enable reading and parsing binary descriptors from non memory mapped flash
+	  (e.g. external flash).
+
+if BINDESC_READ_FLASH
+
+config BINDESC_READ_FLASH_MAX_DATA_SIZE
+	int "Bindesc read flash max data size"
+	range 4 65535
```
Comment (pdgendt) [inline subsys/bindesc/Kconfig:65]: Same

File: scripts/west_commands/bindesc.py (hunk 3 above)
Comment (yonsch) [reply scripts/west_commands/bindesc.py:202]: The magic is 8 bytes long, so it's very unlikely to appear randomly in the image. What about other appearances of the magic, e.g. two images in one file? Well, currently it's not handled. We could add `skip` or `start_offset` arguments, but that deserves more thought. I would say this is very unlikely to happen, as the descriptors are by design placed as close as possible to the start of the image.

**Review by pdgendt — APPROVED** [review]

**Comment by yonsch** [comment]
@Laczen I hope this resolves the security concerns brought up regarding the original PR [here](https://github.com/zephyrproject-rtos/zephyr/pull/54464#issuecomment-1448503697)

**Comment by pdgendt** [comment]
It needs a second approval and @kartben should address his NAK

**Review by kartben — APPROVED** [review]
Sorry for the delay. I think my comments have been addressed and this LGTM :) Thanks for your patience!

**Pull request merged**